### PR TITLE
ASCII block pseudo-device

### PR DESCRIPTION
**Pull request by rm-hull**
image

### CODE REVIEW

**Review by thijstriemstra — CHANGES_REQUESTED** [review]
Can you add a test or two for this emulator?

File: luma/emulator/device.py [hunk 1]
```diff
@@ -310,3 +313,78 @@ def cleanup(self):
...
+class asciiblock(emulator):
+    """
+    Pseudo-device that acts like a physical display, except that it converts
+    the image pixels to display into colored ASCII half-blocks (ASCII code 220,
+    '▄'), where the upper part background is used for one pixel, and the lower
+    part foreground is used for the pixel on the next row. As most terminal
+    display characters are in ratio 2:1, the half-block appears square.
+
+    Inspired by **Command Line Curiosities - Making the Terminal Sing by Hamza Haiken**
+    https://www.youtube.com/watch?v=j5zA5Xi_ph8
```
Comment (thijstriemstra) [inline luma/emulator/device.py]: can you link this instead using


```
`Command Line Curiosities - Making the Terminal Sing by Hamza Haiken <link here>`__
```

File: tests/test_asciiblock.py [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+# See LICENSE.rst for details.
+
+"""
+Tests for :py:class:`luma.emulator.device.asciiblock`.
+"""
+
+import hashlib
+import struct
+import sys
+try:
```
Comment (thijstriemstra) [inline tests/test_asciiblock.py]: @rm-hull we do not support older versions of Python so the `io` version should be enough.

Comment (rm-hull) [reply tests/test_asciiblock.py]: Get some strange unicode encoding problems on py27 using io.StringIO:

```
self = <luma.emulator.device.asciiblock object at 0x10f634350>, cmd = '2J'

    def _CSI(self, cmd):
        """
            Control sequence introducer
            """
>       sys.stdout.write('\x1b[')
E       TypeError: unicode argument expected, got 'str'
```

File: luma/emulator/device.py (hunk 1 above)
Comment (rm-hull) [reply luma/emulator/device.py]: Fixed

File: tests/test_asciiblock.py (hunk 2 above)
Comment (rm-hull) [reply tests/test_asciiblock.py]: So i can use io.BytesIO instead for py27

**Review by thijstriemstra — APPROVED** [review]

**Pull request merged**